UI/EncyclopediaDetailPanel.cpp

```diff
@@ -985,70 +1007,97 @@ void EncyclopediaDetailPanel::HandleSearchTextEntered() {
-        if (dupe_it != already_listed_results.end())
+        if (listed)
             continue;
 
+        // search for partial word matches: searched-for words that appear
+        // in the title text, not necessarily as a complete word
         for (const std::string& word : words_in_search_text) {
             // reject searches in text for words less than 3 characters
             if (word.size() < 3)
                 continue;
-            if (boost::icontains(entry.first, word)) {
-                match_report += entry.second.first;
-                already_listed_results.insert({entry.second.second, entry.first});
+            if (boost::icontains(entry.second.first, word)) {
+                partial_match_report.emplace(entry.second);
+                listed = true;
+                break;
             }
         }
-    }
-
-    // search for matches within article text
-    if (GetOptionsDB().Get<bool>("UI.encyclopedia.search.articles")) {
-        match_report += "\n" + UserString("ENC_SEARCH_ARTICLE_MATCHES") + "\n\n";
-        for (std::multimap<std::string, std::pair<std::string, std::string>>::value_type& entry : all_pedia_entries_list) {
-            std::set<std::pair<std::string, std::string>>::const_iterator dupe_it =
-                already_listed_results.find({entry.second.second, entry.first});
-            if (dupe_it != already_listed_results.end())
-                continue;
+        if (listed || !search_desc)
+            continue;
 
-            EncyclopediaArticle article_entry = GetPediaArticle(entry.second.second);
-            if (boost::icontains(UserString(article_entry.description), search_text)) {
-                match_report += entry.second.first;
-                already_listed_results.insert({entry.second.second, entry.first});
-                break;
-            }
+        // search for matches within article text
+        EncyclopediaArticle article_entry = GetPediaArticle(entry.first);
+        if (article_entry.description.empty()) {
+            // FIXME Articles generated through GetRefreshDetailPanelInfo
+            // will not be found, as no article currently exists.
+            // If resolved, the following should likely log to warn or error
+            TraceLogger() << "No description found: " << entry.first;
+            continue;
         }
+        if (boost::icontains(UserString(article_entry.description), search_text))
+            article_match_report.emplace(entry.second);
     }
 
     // compile list of articles into some dynamically generated search report text
+    std::string match_report = "";
+    bool first_report = true;
+    if (!exact_match_report.empty()) {
+        first_report = false;
+        match_report += UserString("ENC_SEARCH_EXACT_MATCHES") + "\n\n";
+        for (auto&& match : exact_match_report)
+            match_report += match.second;
+    }
+
+    if (!word_match_report.empty()) {
+        if (!first_report)
```

Comment:
this doesn't seem like it will work right if you don't reset `first_report` to true after each heading and double `\n\n`